server/src/main/java/org/opensearch/index/shard/FileUploader.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.index.shard;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.lucene.store.Directory;
+import org.apache.lucene.store.IOContext;
+import org.opensearch.common.CheckedFunction;
+import org.opensearch.common.util.UploadListener;
+import org.opensearch.index.store.RemoteSegmentStoreDirectory;
+
+import java.io.IOException;
+import java.util.Collection;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * This class is a wrapper over the copying of file from local to remote store allowing to decorate the actual copy
+ * method along with adding hooks of code that can be run before, on success and on failure.
+ *
+ * @opensearch.internal
+ */
+public class FileUploader {
+
+    private static final Logger logger = LogManager.getLogger(FileUploader.class);
+
+    private final UploadListener uploadListener;
+
+    private final RemoteSegmentStoreDirectory remoteDirectory;
+
+    private final Directory storeDirectory;
+
+    private final Set<String> excludeFiles;
+
+    private final CheckedFunction<String, String, IOException> checksumProvider;
+
+    public FileUploader(
+        UploadListener uploadListener,
+        RemoteSegmentStoreDirectory remoteDirectory,
+        Directory storeDirectory,
+        Set<String> excludeFiles,
+        CheckedFunction<String, String, IOException> checksumProvider
+    ) {
+        this.uploadListener = uploadListener;
+        this.remoteDirectory = remoteDirectory;
+        this.storeDirectory = storeDirectory;
+        this.excludeFiles = excludeFiles;
+        this.checksumProvider = checksumProvider;
+    }
+
+    /**
+     * Calling this method will filter out files that need to be skipped and call
+     * {@link RemoteSegmentStoreDirectory#copyFilesFrom}
+     *
+     * @param files The files that need to be uploaded
+     * @return A boolean for whether all files were successful or not
+     * @throws Exception when the underlying upload fails
+     */
+    public boolean uploadFiles(Collection<String> files) throws Exception {
+        Collection<String> filteredFiles = files.stream().filter(file -> !skipUpload(file)).collect(Collectors.toList());
+        return remoteDirectory.copyFilesFrom(storeDirectory, filteredFiles, IOContext.DEFAULT, uploadListener);
+    }
+
+    /**
```

Comment:
The concern I have with this separate wrapper interface is it can easily be by-passed by someone who directly interacts with remote directory interface, in which case it will be hard to maintain sync.